src/Worker.Extensions.DurableTask/OrchestrationInputConverter.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.Worker.Converters;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.DurableTask;
+
+/// <summary>
+/// Binds the input for an orchestration, if an input parameter is present.
+/// </summary>
+internal class OrchestrationInputConverter : IInputConverter
+{
+    /// <summary>
+    /// The key we cache the prepared orchestration input in <see cref="FunctionContext.Items" />.
+    /// </summary>
+    private const string OrchestrationInputKey = "__orchestrationInput__";
+
+    /// <summary>
+    /// Gets the input context for the current orchestration invocation. This context can be later
+    /// used to "seed" the input for later conversion.
+    /// </summary>
+    /// <param name="context">The function context.</param>
+    /// <returns>The input context.</returns>
+    public static InputContext GetInputContext(FunctionContext context)
+    {
+        if (context is null)
+        {
+            throw new ArgumentNullException(nameof(context));
+        }
+
+        FunctionDefinition definition = context.FunctionDefinition;
+        foreach (FunctionParameter parameter in definition.Parameters)
+        {
+            if (!IgnoredInputType(parameter.Type)
+                && !definition.InputBindings.ContainsKey(parameter.Name)
+                && !definition.OutputBindings.ContainsKey(parameter.Name))
+            {
+                // We take the first parameter we encounter which is not an explicitly ignored type
+                // and has exactly 0 bindings (in or out) as our input candidate.
+                return new(parameter.Type, context);
+            }
```

Comment:
For safety - should we consider not short-circuiting this loop so that we raise an exception if more than 1 candidate is found? I would want us to avoid "silent" assumption violations if possible.

Reply:
The functions pipeline will handle throwing for us if there are unbound parameters at the end of input binding. I don't think this should throw in this case because sometimes it _could_ be valid as well - there might be another registered input converter that will handle the additional candidates.

Albeit this is all kind of fragile as the behavior here can change depending on what extensions the customer is using and in what order there are registered. Unfortunately, we cannot control for that, but I think this feature is worth it even with that potential flaw.

Reply:
I agree this is a little fragile. My worry is that the Functions team could introduce a new type of input that we didn't account for and then we'll need to react by adding it to `IgnoredInputType`. Have you considered a design where we use a custom attribute to mark a particular parameter as being an orchestration/activity input?

Reply:
A specific parameter is doable for sure, but it is slightly superfluous I find: `public async Task MyOrchestration([OrchestrationTrigger] TaskOrchestrationContext context, [OrchestrationInput] TInput input)`.

This fragility I mentioned is something already part of input binding as a whole and not unique to this scenario. Much like DI, the user can mess with the DI container and cause unexpected behavior. Here the user can also mess with input converters and cause all sorts of unexpected behavior. If users have an issue with this, the old way of `.GetInput<T>` will still be available to them.